### PR TITLE
Hammer: implement piece storage

### PR DESCRIPTION
**Pull request by egonelbre**

### CODE REVIEW

File: storagenode/pieces/readwrite.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	w := &Writer{}
+	w.buf = *bufio.NewWriterSize(blob, bufferSize)
+	w.blob = blob
+	w.hash = pkcrypto.NewHash()
+	return w, nil
+}
+
+// Write writes data to the blob and calculates the hash.
+func (w *Writer) Write(data []byte) (int, error) {
+	_, err := w.hash.Write(data)
```
Comment (kaloyan-raev) [inline storagenode/pieces/readwrite.go]: Why are we adding the data to the hash first? Should we add only those bytes that were _actually_ written to buf _after_ writing them successfully?

File: storagenode/pieces/readwrite.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	if err != nil {
+		return 0, Error.Wrap(err)
+	}
+	n, err := w.buf.Write(data)
```
Comment (mniewrzal) [inline storagenode/pieces/readwrite.go]: Is it possible that hash will write more bytes than `buf`?

Comment (egonelbre) [reply storagenode/pieces/readwrite.go]: ah, right

Comment (mniewrzal) [reply storagenode/pieces/readwrite.go]: Kaloyan better addressed my concerns :)

File: storagenode/pieces/readwrite.go (hunk 1 above)
Comment (egonelbre) [reply storagenode/pieces/readwrite.go]: yes, fixed

File: uplink/piecestore/upload.go [hunk 3]
```diff
@@ -66,8 +66,10 @@ func (client *Upload) Write(data []byte) (written int, _ error) {
 		return 0, ErrProtocol.Wrap(client.sendError)
 	}
 
-	// hash the content so far
-	_, _ = client.hash.Write(data) // guaranteed not to return error
+	fullData := data
+	defer func() {
+		_, _ = client.hash.Write(fullData[:written]) // guaranteed not to return error
```
Comment (mniewrzal) [inline uplink/piecestore/upload.go]: You can use `pkcrypto.NewHash()` for this `client.hash` in constructor too I guess.

**Pull request merged**